### PR TITLE
feat: add support for different type of dependencies

### PR DESCRIPTION
**Pull request by lili2311**
- [x] Tests written and linted [ℹ︎]()
- [x] Commit history is tidy [ℹ︎]()

### What this does

Added more tests for:
- devDependencies only are present
- Same dependency appears twice (once in dependencies and once more in devDependencies)
- empty dependencies and varying the `--dev` flag
### Notes for the reviewer
Run this locally
### More information

- [SC-5985](https://snyksec.atlassian.net/browse/SC-5985)

### CODE REVIEW

File: lib/index.ts [hunk 1]
```diff
@@ -82,9 +94,16 @@ async function buildSubTreeRecursive(dep: string, depKeys: string[], lockFile: o
     depSubTree.depType = deps[dep].dev ? DepType.dev : DepType.prod;
     // repeat the process for dependencies of looked-up dep
     const newDeps = deps[dep].requires ? Object.keys(deps[dep].requires) : [];
-    await Promise.all(newDeps.map(async (subDep) => {
-      depSubTree.dependencies[subDep] = await buildSubTreeRecursive(subDep, [...depKeys, dep], lockFile);
-    }));
+
+    await Promise.all(
+      newDeps
```
[the remark this answers is not in the document]
Comment (lili2311) [reply lib/index.ts]: Let's set this up in linting, I believe this is autoformatted based on our current rules.

File: lib/index.ts [hunk 2]
```diff
@@ -44,23 +41,38 @@ async function buildDepTree(targetFileRaw: string, lockFileRaw: string, includeD
...
+
+function getTopLevelDeps(targetFile, includeDev) {
+  let topLevelDeps = targetFile.dependencies;
+
   if (includeDev && targetFile.devDependencies) {
-    const topLevelDevDeps = Object.keys(targetFile.devDependencies);
-    await Promise.all(topLevelDevDeps.map(async (dep) => {
-      depTree.dependencies[dep] = await buildSubTreeRecursive(dep, [], lockFile);
-    }));
+    topLevelDeps = {
```
[the remark this answers is not in the document]
Comment (lili2311) [reply lib/index.ts]: Updated for uniqueness.

File: lib/index.ts [hunk 3]
```diff
@@ -82,9 +104,12 @@ async function buildSubTreeRecursive(dep: string, depKeys: string[], lockFile: o
     depSubTree.depType = deps[dep].dev ? DepType.dev : DepType.prod;
     // repeat the process for dependencies of looked-up dep
     const newDeps = deps[dep].requires ? Object.keys(deps[dep].requires) : [];
+
     await Promise.all(newDeps.map(async (subDep) => {
-      depSubTree.dependencies[subDep] = await buildSubTreeRecursive(subDep, [...depKeys, dep], lockFile);
-    }));
+      depSubTree.dependencies[subDep] = await buildSubTreeRecursive(subDep,
```
Comment (lili2311) [inline lib/index.ts]: I think our allowed line length is too long, can we consider something smaller than 120?  Perhaps 100?

File: lib/index.ts [hunk 4]
```diff
@@ -18,6 +18,16 @@ interface PkgTree {
   hasDevDependencies?: boolean;
 }
 
+interface TargetFile {
```
Comment (miiila) [inline lib/index.ts]: 😍

File: lib/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (miiila) [reply lib/index.ts]: I am fine with the current solution for now, but I'd like to mention one thing:

By 
```
topLevelDeps = {
  ...topLevelDeps,
  ...targetFile.devDependencies,
}
```
big new object with all the nested subobjects is created, while we are interested in the flat array of keys only. So it's unnecessary memory overhead. Although it won't have any real implications (the most probably), it's a good practice to not use more memory, than you really need.

File: lib/index.ts (hunk 3 above)
Comment (miiila) [reply lib/index.ts]: I am fine with 100.

**Pull request merged**